### PR TITLE
Conformance rules page

### PR DESCRIPTION
**Pull request by pp-mo**
This is an attempt to provide a reasonably watertight set of "conformance rules" for UGRID.
( Draft build I have up  : https://ugrid-conventions.readthedocs.io/en/conformance/conformance/ )

This was identified as a desirable to progress #52.
Also mentioned in more detail to in the CF discussion https://github.com/cf-convention/cf-conventions/issues/153 : I think the most recent summary/agreement is that stated in [this @hrajagers comment](https://github.com/cf-convention/cf-conventions/issues/153#issuecomment-703858946)

My own particular interest is that I am developing some code for UGRID compliance _checking_.
For that, I think it is highly desirable to have a reference document for warning messages to refer ***to*** (as the cf-checker utility does).
So ideally, we would publish it here : then it would be public, agreed, "official" and maintainable.

Having said all that, I doubt we will be in hurry to merge this, because my approach is probably a bit narrow, and certainly rather "opinionated".
What I **_am_** hoping for, at least, is some informed overview and comment.
So... a handy checklist of some aspects that I (already) think might prove contentious...
  * I have styled it as a list of separate short statements, each with an ID code and a linkable anchor
  * I have added a range of advisory "recommendations", virtually all of which I made up myself (!)
  * I am allowing "nodes-only" meshes, with `topology_dimension` = 0
  * I have (so far) omitted all the 3d/volume facilities, firstly because it is huge simplification, but also
    as it has lately been suggested [we may drop these altogether in the current form](https://github.com/ugrid-conventions/ugrid-conventions/issues/53#issuecomment-831906537).
     * To resolve that, it will be essential to _either_ remove the 3d aspects from the current main page, _or_ add supporting rules for 3d here.  I'm hoping for more informed feedback to make a decision on this.

### CODE REVIEW

**Comment by pp-mo** [comment]
Ok, I finally got around to some serious work on https://github.com/pp-mo/ugrid-checks/tree/main/lib/ugrid_checks
( where I now have a checker working + almost complete )
As predicted, that prompted some changes here, which I just pushed.

I'm now confident that this set of rules is workable (from the checker POV), so I took it out of draft.

My remaining concern here is to clearly label this "DRAFT", as @ChrisBarker-NOAA suggests.
I thought the nicest way would be a watermarking, so had a quick try at implementing the solution described [here on StackOverflow](https://stackoverflow.com/a/40854345/2615050), but my understanding of web tech is very thin and I have so far failed (!)
It's possible that mkdocs is in the way, but more likely IMHO that I don't understand CSS.
*Any ideas ?*
Anyway, I am back at my day job next week, where there is someone I can ask about these things, so I will try to pursue the watermarking idea that way.

File: src/conformance.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+[TOC]
+
+## Introduction
+
+This information is intended to follow the practice established by the
+[CF conformance pages][cf_conformance], as attached to the [CF conventions][cf_conventions]
+
+Each reference statement is provided with a short code :
+"Rxxx" for requirements and "Axxx" for advisory recommendations,
+in the manner of the [Flake8 error and warning codes][flake8_codes], and the [PyCodeStyle ones][pycodestyle_codes].
```
[the remark this answers is not in the document]
Comment (ChrisBarker-NOAA) [reply src/conformance.md:12]: Sound good to me.

**Review by ChrisBarker-NOAA — COMMENTED** [review]
A Watermark would be nice, but I have no idea how to do it, and I dont hink it's worth all that much effort.

I think a big:

"THIS IS A DRAFT" at the top of the page (with maybe a touch of explanatory text)  is totally fine.

**Comment by pp-mo** [comment]
Ok, I think I'm happy with this now.

FYI I tried a bunch of ideas for 'watermarking', but none of interacted predictably with markdown/mkdocs, so I've gone with a simpler approach which totally avoids CSS (as you suggested @ChrisBarker-NOAA !)

**Comment by pp-mo** [comment]
> @ChrisBarker-NOAA Maybe you could put a big "DRAFT" note at the top, and then we could merge it. It would be nice to have it published for easier review and comment.

Regarding that, if you merge does this automatically appear on the website itself ?

**Comment by pp-mo** [comment]
> does this automatically appear on the website itself ?

OK I think I now understand that it does, via Travis ?

**Comment by ChrisBarker-NOAA** [comment]
yes, I think they auto-build -- though I don't know that it's been tested in a while, so we'll see :-)

@core: I think it's a good idea to merge this, but I'd like at least one other person to agree. If any of you do agree, go ahead and merge.

**Comment by pp-mo** [comment]
Hi @ChrisBarker-NOAA 
just FYI, I have just now published a installable "beta" release of my UGRID conformance checker utility based on these rules : https://github.com/pp-mo/ugrid-checks 

It is pretty functional now, and I wanted to share it for internal use.  
So I have now made it buildable+installable, added basic usage instructions in the README, tagged v0.1 and pushed to PyPI .
( I also particularly wanted to get out something useable,  so that I can concentrate on other matters again for a while. )
We can maybe discuss that "living" in this organisation, too.  I hope I haven't made dev tooling too involved for collaborators, but if so I'd be open to simplifying it.

As it made sense for the README usage instructions to refer to a list of conformance rules, _just for now_ I have linked that account to [a private RTD build including the current conformance page](https://ugrid-conventions.readthedocs.io/en/conformance/conformance/).
So, I hope that doesn't prejudice the principle that it is not agreed here yet !!

In view of @davidhassell recent progress on https://github.com/cf-convention/cf-conventions/pull/353, it would be nice to come to some at-least-interim agreement on the conformance content here.
It **_also_** would be really good to resolve the [question of the 3d/volume support](https://github.com/ugrid-conventions/ugrid-conventions/issues/53#issuecomment-832138564), since that is another obstacle to adoption in CF.
I think I should be able to find some contacts here at the MetOffice who could comment on a possible "function space" encoding, as mentioned in relation to Fluidity/ICOM in that same discussion.  For us that is still only a vague future possibility, I think, but I will see if anyone can be persuaded to comment!

Does it seem like we can address these things now, or do you see a better (future) time coming ?

**Comment by ChrisBarker-NOAA** [comment]
@pp-mo: This looks really great! I've taken a quick look at your code, and it looks really nicely structured. tested, etc.

I haven't tried it much -- only one small example, file, but it worked well on that one :-)

Coverage of check,py is %80 - not perfect, but pretty good.
```
Name                                                                        Stmts   Miss  Cover
-----------------------------------------------------------------------------------------------
/Users/chris.barker/Temp/ugrid-checks/lib/ugrid_checks/__init__.py              4      0   100%
/Users/chris.barker/Temp/ugrid-checks/lib/ugrid_checks/check.py               803    166    79%
/Users/chris.barker/Temp/ugrid-checks/lib/ugrid_checks/nc_dataset_scan.py      64      6    91%
/Users/chris.barker/Temp/ugrid-checks/lib/ugrid_checks/scan_utils.py           27      0   100%
/Users/chris.barker/Temp/ugrid-checks/lib/ugrid_checks/ugrid_logger.py         70     10    86%
__init__.py                                                                    72      0   100%
check/__init__.py                                                               0      0   100%
check/test_check_dataset.py                                                   746      1    99%
nc_dataset_scan/__init__.py                                                     0      0   100%
nc_dataset_scan/test_scan_dataset.py                                           57      0   100%
-----------------------------------------------------------------------------------------------
TOTAL                                                                        1843    183    90%
```
 
As to where to go now:

If you think it's ready -- maybe we should make a conda=forge package for it - a lot of u use conda for netcdf works, etc.

>We can maybe discuss that "living" in this organization, too.

I support that -- makes all the sense in the world

> I hope I haven't made dev tooling too involved for collaborators, but if so I'd be open to simplifying it.

I didn't see anything scary :-)

>  it would be nice to come to some at-least-interim agreement on the conformance content here.

The challenge is folks finding the time to review it -- I gave it a quick once over and didn't see anything to complain about.

So maybe we can ping the group ( I wonder if the old email list is still functional?) to ask for review.

But I'm a believer that something mostly done is better than nothing :-) -- after all, the UGRID spec was a draft for a very long time. So even if we can't get much review, I'd be happy to see it merged in as a draft.

> It also would be really good to resolve the [question of the 3d/volume support](https://github.com/ugrid-conventions/ugrid-conventions/issues/53#issuecomment-832138564), since that is another obstacle to adoption in CF.

Well, I don't know of anyone trying to use UGRID for 3-D volumes, and I haven't seen anyone step up to tackle it, so I think we can simply say that 3D support is tentative and not CF compliant.

> Does it seem like we can address these things now, or do you see a better (future) time coming ?

Now is better than never -- let's do it!

(barring any objections from anyone else in the group)

@hrajagers, @rsignell-usgs, @ocefpaf : any thoughts?

**Comment by pp-mo** [comment]
@ChrisBarker-NOAA thanks so much for your attention + thoughtful discussion.

> If you think it's ready -- maybe we should make a conda=forge package for it

I think I can easily sort that out.  We do use conda use here mostly anyway, but it seemed that  pip/PyPI was jusy a bit quicker for minimal effort to get an initial release out.  I noted that the structure of conda recipes has shifted a bit since I last handled  one (only a few months ago!), but I am lucky to have other people in my group who do have a  good up-to-date knowledge, so I hope to progress that soon.
TBH I would also like to fix a known small bug first, though (see below)

---
> Coverage of check,py is %80 

Yes indeed !  I have been doing occasional coverage checks, and I _think_ all the actual checking code is well covered, but other less essential aspects are still a bit short of testing.  
Notably, since the version 0.1 I find I missed something quite important in the 'summary' output https://github.com/pp-mo/ugrid-checks/issues/20  -- that feature was a late addition, which I did not add tests for, so that just goes to show.  I hope to fix it quite soon.

**Do you feel that a coverage check should be built into CI testing ?**
I'm never sure how I feel about that myself.  It's hard to set pass criteria and can sometimes seem like a bit of a pointless box-ticking exercise.

---
> I think we can simply say that 3D support is tentative and not CF compliant.

OK that is encouraging.  And I think it makes great sense, in the spirit of your
> something mostly done is better than nothing :-)

**Comment by pp-mo** [comment]
Hi @ChrisBarker-NOAA quick update.
I have now found time to fix the bugs I identified with 0.1, improve testing, and push that out as [version 0.1.1](https://github.com/pp-mo/ugrid-checks/releases/tag/v0.1.1)

The coverage is now much higher, and for sure it was worth writing proper tests because I found at least 2 more bugs with the structure-report :  ( It really was a bit last-minute and not ready )

What may interest you more ... with a little help from @bjlittle it is now also **available on conda-forge** :bell: ✔️ 
https://github.com/conda-forge/ugrid-checks-feedstock

----
However ... I have also in this process learnt that there is some prior art I was not previously aware of, 
in https://github.com/ioos/cc-plugin-ugrid

So, I will next be engaging with that project and seeing where this takes us.
the connection was made by @ocefpaf in [this discussion](https://github.com/conda-forge/staged-recipes/pull/18116#pullrequestreview-889644800) 

I likewise been asked by a couple of colleagues whether this would eventually be subsumed in [cf-checker](https://github.com/cedadev/cf-checker).
I must say, I still don't really have much of a view on that.   It seems practically some way off possibility at present, which is why I decided not to start there.  **Do you have any views on that ?**

**Comment by ChrisBarker-NOAA** [comment]
As for versioning -- the convention version is a version of the convention itself, not the document. So any changes/updated to the document(s) shouldn't change the version number of the convention.

That being said, it there's something to be said for tracking the version of the documents -- maybe a separate version number, or use the patch version? (e.g. 1.0.3 for a third update to the docs that don't change the convention)

**Comment by ChrisBarker-NOAA** [comment]
@hrajagers, @rsignell-usgs, @ocefpaf, or anyone else:

I think we should merge this -- a draft is better than nothing, particularly given that UGRID may be officially mentioned in CF soon (https://github.com/cf-convention/cf-conventions/issues/153)

But I'd like at least one other voice of consent :-)

**Review by ocefpaf — APPROVED** [review]

**Comment by davidhassell** [comment]
Hello,

UGRID is all set to be incorporated in to CF-1.11, with the only barrier being the creation of the UGRID conformance document. @ChrisBarker-NOAA [suggested](https://github.com/ugrid-conventions/ugrid-conventions/pull/54#issuecomment-1219837348) that merging this PR to publish the conformance rules  as a draft would be fine for this purpose, and I agree. 

Is this something that you're all happy to do, do you think?

Thanks,
David

**Comment by pp-mo** [comment]
> Is this something that you're all happy to do, do you think?

For reference, it does contain the slightly radical suggestion to omit the full 3D aspect.
On the other hand, IMHO it can be accepted "as draft", as already suggested.

**Comment by ChrisBarker-NOAA** [comment]
@pp-mo wrote:

"For reference, it does contain the slightly radical suggestion to omit the full 3D aspect."

As far as I could see -- it's only omitting 3D from the conformance rules -- which doesn't seem radical to me.   Or does that mean a 3D mesh would be non-conforming by these rules, as opposed to conformance-unknown. 

If it would b enon-conforming, then we could call these conformance rules for < 3D meshes, and be done for now.

**Comment by davidhassell** [comment]
Hi - remembering a [comment from @hrajagers](https://github.com/ugrid-conventions/ugrid-conventions/issues/53#issuecomment-831906537), 3D meshes are currently not fully defined, because the ordering of the nodes is ambiguous, and the suggestion was to fix the numbering, or drop 3D meshes until they're needed. The latter approach gained some support ... so not including them in the conformance doc seems reasonable to me (although I have no strong opinion, nor sway, here!)

**Comment by ChrisBarker-NOAA** [comment]
Ideally, we’d make the experimental nature of the 3D meshes more clear, but in the meantime, I think this is fine.-CHBChristopher Barker, Ph.D.OceanographerEmergency Response DivisionNOAA/NOS/OR&R            (206) 526-6959   voice7600 Sand Point Way NE   (206) 526-6329   faxSeattle, WA  98115       (206) 526-6317   main receptionOn Dec 8, 2022, at 11:53 AM, David Hassell ***@***.***> wrote:﻿
Hi - remembering a comment from @hrajagers, 3D meshes are currently not fully defined, because the ordering of the nodes is ambiguous, and the suggestion was to fix the numbering, or drop 3D meshes until they're needed. The latter approach gained some support ... so not including them in the conformance doc seems reasonable to me (although I have no strong opinion, nor sway, here!)

File: src/conformance.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+For any mesh variable ..
+
+- <a id="R101" href="#R101">R101</a> it must have a `cf_role` attribute,
+    - <a id="R102" href="#R102">R102</a> with the value `"mesh_topology"`
+- <a id="R103" href="#R103">R103</a> it must have a `topology_dimension` attribute, 
+    - <a id="R104" href="#R104">R104</a> whose value is an integer between 0 and 2. [^4] [^5]
+- all mesh connectivity and coordinate attributes must be varlists, that is ..
+    - <a id="R105" href="#R105">R105</a> they are strings consisting of space-separated, valid netcdf variable names, and
+    - <a id="R106" href="#R106">R106</a> the variables named all exist in the dataset
+- <a id="R107" href="#R107">R107</a> all mesh coordinate attributes must contain a single variable name
```
Comment (pp-mo) [inline src/conformance.md:81]:
```suggestion
- <a id="R107" href="#R107">R107</a> all mesh connectivity attributes must contain a single variable name
```
Typo !

File: src/conformance.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+  * UGRID v1.0 
+  * [CF version 1.9][cf_current_version]
+
+---
+
+[TOC]
+
+## Preamble : terms and definitions
+- A ***mesh*** is a variable with a `cf_role` attribute of `"mesh_topology"`. [^2]
```
Comment (pp-mo) [inline src/conformance.md:34]:
```suggestion
- A ***mesh variable*** is a variable with a `cf_role` attribute of `"mesh_topology"`. [^2]
```
I think more consistent with usage below

File: src/conformance.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+Each reference statement is provided with a short code :
+"Rxxx" for requirements and "Axxx" for advisory recommendations,
+in the manner of the [Flake8 error and warning codes][flake8_codes], and the [PyCodeStyle ones][pycodestyle_codes].
+<br>Future editions will maintain these reference numbers, but possibly add new ones.  They can be used as definitive
+references by a conformance-checker utility. [^10]
+
+**This version** is consistent with :
+
+  * UGRID v1.0 
+  * [CF version 1.9][cf_current_version]
```
Comment (pp-mo) [inline src/conformance.md:27]: Update to 1.10 ?

File: src/conformance.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+   However, if the connectivity is correctly formed, then its `cf_role` value will always be the same as this  -- cf:[R303](#R303).
+
+[^14]: in the present [latest CF version][cf_current_version] the only `cf_role` terms defined are the discrete-geometry
+  types : `timeseries_id`, `profile_id` and `trajectory_id`
+
+
+[volume_status]: https://github.com/ugrid-conventions/ugrid-conventions/issues/53#issuecomment-832138564
+[cf_conformance]: http://cfconventions.org/cf-conventions/conformance.html "CF Conventions Conformance pages"
+[cf_conventions]: https://cfconventions.org/ "CF metadata conventions"
+[cf_current_version]: https://cfconventions.org/Data/cf-conventions/cf-conventions-1.9/cf-conventions.html "CF version 1.9"
```
Comment (pp-mo) [inline src/conformance.md:289]: Update to 1.10 ?

**Review by pp-mo — COMMENTED** [review]
Just reviewing this for the first time in a while, spotted a few things that can be improved.
But N.B. cannot approve or require changes to your own PR !!

File: src/conformance.md (hunk 5 above)
Comment (ChrisBarker-NOAA) [reply src/conformance.md:289]: I think if you make the changes to your own fork, the will automatically show up in the PR. But I'll take a quick look right now, and at least approve the typo fixes :-)

File: src/conformance.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+      (see [below](#edge-dimension) for how this is determined)
+    - if there is a `face_node_connectivity`, then there is a **face** dimension
+      (see [below](#face-dimension) for how this is determined)
+    - if there is a `boundary_node_connectivity`, then there is a **boundary** dimension
+      (see [below](#boundary-dimension) for how this is determined)
+- the ***element dimension*** of a mesh connectivity or mesh coordinate is that one of its dimensions
+  which is an element dimension of its parent mesh.
+  N.B. there can only be one : see [R201](#R201) / [R304](#R304)
+- A ***location index set*** is a variable with a `cf_role` attribute of `"location_index_set"` [^2]
+- A ***mesh data variable*** is one with either a `mesh` or a `location_index_set` attribute.
```
[the remark this answers is not in the document]
Comment (ChrisBarker-NOAA) [reply src/conformance.md:67]: I'm not sure CF is relevant here. -- how about:

- A ***mesh data variable*** is a data variable with either a `mesh` or a `location_index_set` attribute.

File: src/conformance.md (hunk 4 above)
Comment (ChrisBarker-NOAA) [reply src/conformance.md:27]: Hmm -- UGRID just got added to CF, specifically with 1.0 -- so let's keep 1.0

**Review by ChrisBarker-NOAA — APPROVED** [review]

File: src/conformance.md (hunk 5 above)
Comment (pp-mo) [reply src/conformance.md:289]: I guess this should now say "1.11".
Will fix ...

File: src/conformance.md (hunk 4 above)
Comment (pp-mo) [inline src/conformance.md:27]:
```suggestion
  * [CF version 1.11][cf_current_version]
```

**Pull request merged**